File: tests/fixtures/python/clean_code_uniformity.py
from typing import Iterator

Ledger = dict[str, int]


def apply(ledger: Ledger, entries: list[tuple[str, int]]) -> int:
    total = 0
    for account, amount in entries:
        if not account:
            raise ValueError("entry with no account")
        ledger[account] = ledger.get(account, 0) + amount
        total += amount
    return total


def balance(ledger: Ledger, account: str) -> int:
    return ledger.get(account, 0)


# Largest-first, because a settlement run that stops early should still have
# cleared the balances that matter most.
def settle(ledger: Ledger, budget: int) -> list[str]:
    owed = sorted(
        ((account, amount) for account, amount in ledger.items() if amount > 0),
        key=lambda pair: (-pair[1], pair[0]),
    )

    remaining = budget
    settled = []
    for account, amount in owed:
        if amount > remaining:
            continue
        remaining -= amount
        ledger[account] = 0
        settled.append(account)
    return settled


def merge(left: Ledger, right: Ledger) -> Ledger:
    out = dict(left)
    for account, amount in right.items():
        out[account] = out.get(account, 0) + amount
    return out


def nonzero(ledger: Ledger) -> Iterator[tuple[str, int]]:
    return ((account, amount) for account, amount in ledger.items() if amount)


def describe(ledger: Ledger) -> str:
    parts = [f"{account}={amount}" for account, amount in nonzero(ledger)]
    if not parts:
        return "empty ledger"
    return ", ".join(parts)


def rebase(ledger: Ledger, floor_value: int) -> Ledger:
    out: Ledger = {}
    for account, amount in ledger.items():
        out[account] = amount if amount >= floor_value else floor_value
    return out


def largest(ledger: Ledger) -> str | None:
    best = None
    best_amount = float("-inf")
    for account, amount in ledger.items():
        if amount > best_amount:
            best = account
            best_amount = amount
    return best


def split_by_sign(ledger: Ledger) -> tuple[Ledger, Ledger]:
    credits: Ledger = {}
    debits: Ledger = {}
    for account, amount in ledger.items():
        target = credits if amount >= 0 else debits
        target[account] = amount
    return credits, debits


def totals(ledger: Ledger) -> tuple[int, int]:
    credits, debits = split_by_sign(ledger)
    return sum(credits.values()), sum(debits.values())

File: tests/fixtures/typescript/clean_code_uniformity.ts
export type Ledger = Map<string, number>;

export type Entry = { account: string; amount: number };

export function apply(ledger: Ledger, entries: Entry[]): number {
  let total = 0;
  for (const { account, amount } of entries) {
    if (account.length === 0) {
      throw new Error("entry with no account");
    }
    if (!Number.isFinite(amount)) {
      throw new Error(`non-finite amount on ${account}`);
    }
    ledger.set(account, (ledger.get(account) ?? 0) + amount);
    total += amount;
  }
  return total;
}

export function balance(ledger: Ledger, account: string): number {
  return ledger.get(account) ?? 0;
}

// Largest-first, because a settlement run that stops early should still have
// cleared the balances that matter most.
export function settle(ledger: Ledger, budget: number): string[] {
  const owed = [...ledger.entries()]
    .filter(([, amount]) => amount > 0)
    .sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0]));

  let remaining = budget;
  const settled: string[] = [];
  for (const [account, amount] of owed) {
    if (amount > remaining) {
      continue;
    }
    remaining -= amount;
    ledger.set(account, 0);
    settled.push(account);
  }
  return settled;
}

export function merge(left: Ledger, right: Ledger): Ledger {
  const out = new Map(left);
  for (const [account, amount] of right) {
    out.set(account, (out.get(account) ?? 0) + amount);
  }
  return out;
}

export function* nonzero(ledger: Ledger): Generator<[string, number]> {
  for (const pair of ledger) {
    if (pair[1] !== 0) {
      yield pair;
    }
  }
}

export function describe(ledger: Ledger): string {
  const parts: string[] = [];
  for (const [account, amount] of nonzero(ledger)) {
    parts.push(`${account}=${amount}`);
  }
  return parts.length === 0 ? "empty ledger" : parts.join(", ");
}

export function rebase(ledger: Ledger, floorValue: number): Ledger {
  const out: Ledger = new Map();
  for (const [account, amount] of ledger) {
    out.set(account, Math.max(amount, floorValue));
  }
  return out;
}

export function largest(ledger: Ledger): string | undefined {
  let best: string | undefined;
  let bestAmount = Number.NEGATIVE_INFINITY;
  for (const [account, amount] of ledger) {
    if (amount > bestAmount) {
      best = account;
      bestAmount = amount;
    }
  }
  return best;
}
